src/os.rs

```diff
@@ -576,27 +576,60 @@ mod imp {
 
 #[cfg(all(target_arch = "wasm32", not(target_os = "emscripten")))]
 mod imp {
-    use std::io;
+    use stdweb::unstable::TryInto;
+    use stdweb::web::error::Error as WebError;
+    use {Error, ErrorKind};
 
     #[derive(Debug)]
     pub struct OsRng;
 
     impl OsRng {
         pub fn new() -> Result<OsRng, Error> {
-            Err(Error::new(ErrorKind::Unavailable, "not supported on WASM"))
+            Ok(OsRng)
         }
+
         pub fn try_fill_bytes(&mut self, v: &mut [u8]) -> Result<(), Error> {
-            Err(Error::new(ErrorKind::Unavailable, "not supported on WASM"))
+            let len = v.len() as u32;
+            let ptr = v.as_mut_ptr() as i32;
+
+            let result = js! {
+                let len = @{ len };
+                let ptr = @{ ptr };
+
+                try {
+                    try {
+                        let array = new Uint8Array(len);
+                        window.crypto.getRandomValues(array);
+                        HEAPU8.set(array, ptr);
+                    } catch(err) {
+                        if (err instanceof ReferenceError) {
+                            let bytes = require("crypto").randomBytes(len);
+                            HEAPU8.set(new Uint8Array(bytes), ptr);
+                        } else {
+                            throw err;
+                        }
+                    }
+
+                    return { success: true };
+                } catch(err) {
+                    return { success: false, error: err };
+                }
+            };
+
+            if js!{ return @{ result.as_ref() }.success } == true {
+                Ok(())
+            } else {
+                let err: WebError = js!{ return @{ result }.error }.try_into().unwrap();
+                Err(Error::with_cause(ErrorKind::Other, "WASM Error", err))
```

Comment:
You should probably give a `Unavailable` error if `window.crypto.getRandomValues` is not available (can that happen, or does every implementation of webassembly support this?). And do you know whenwe can get a `QuotaExceededError`?

b.t.w. `Other` just got renamed to `Unknown`.

Reply:
No, to `Unexpected`. But I agree, it doesn't look like this is recoverable so it should be `Unavailable`.

Reply:
Found an answer: https://w3c.github.io/webcrypto/Overview.html#dfn-Crypto-method-getRandomValues. `QuotaExceededError` is returned "If the byteLength of array is greater than 65536". We maybe want to break up larger request, like the implementations of `OsRng` on other systems (although I would say an error is reasonable...)